### PR TITLE
feat: Implement new authentication framework design

### PR DESCRIPTION
**Pull request by padamstx**
Fixes: https://github.ibm.com/arf/planning-sdk-squad/issues/981

This PR contains lots of changes in the java core to implement the new strategic authentication design, located [here](https://github.ibm.com/CloudEngineering/openapi-sdkgen/wiki/Core-SDK-Authentication-Design)

This is a fairly large PR with lots of moving parts, so I'd be happy to do a "pre-review" review to go over the main parts.

### CODE REVIEW

**Comment by padamstx** [comment]
I cc'd the other core owners mainly for awareness, as the implementations of each authenticator across the different language-specific cores should be as consistent as we can make them.

**Comment by germanattanasio** [comment]
@padamstx can you write as a comment here what would users do to use this in all our environments?

**Comment by padamstx** [comment]
> @padamstx can you write as a comment here what would users do to use this in all our environments?

Good idea, but I'll add the usage examples to the design wiki that @dpopp07 introduced yesterday instead of this PR.

**Comment by lpatino10** [comment]
The design document looks really good to me! Thanks @dpopp07 and @padamstx.

Taking a look at this now 🔍

**Comment by germanattanasio** [comment]
I think the examples are missing the service url which is different than the authentication url

File: src/main/java/com/ibm/cloud/sdk/core/util/CredentialUtils.java [hunk 1]
```diff
@@ -462,4 +400,149 @@ private static ServiceCredentials setCredentialFields(String serviceName, List<S
...
     return serviceCredentials;
   }
+
+  /**
+   * Returns a Map containing properties found within the credential file that are associated with the
+   * specified cloud service.
+   * @param serviceName the name of the cloud service whose properties should be loaded
+   * @return a Map containing the properties
+   */
+  public static Map<String, String> getFileCredentialsAsMap(String serviceName) {
```
Comment (lpatino10) [inline src/main/java/com/ibm/cloud/sdk/core/util/CredentialUtils.java]: It looks like there's actually a lot more we can clean up in this file.
- `getFileCredentialsAsMap()` is almost identical to the old `getFileCredentials()`, other than the return type. I think we can just get rid of the old method.
- `getCredentialsFromVcap()` is also pretty redundant with the addition of `getVcapCredentialsAsMap()`. I think we can refactor the latter method to just try and get the individual fields directly. Something like:

    ```java
    public static Map<String, String> getVcapCredentialsAsMap(String serviceName) {
      Map<String, String> props = new HashMap<>();
      addToMap(props, Authenticator.PROPNAME_USERNAME, getVcapValue(serviceName, Authenticator.PROPNAME_USERNAME));
      addToMap(props, Authenticator.PROPNAME_PASSWORD, getVcapValue(serviceName, Authenticator.PROPNAME_PASSWORD));
      // etc...
    }
    ```
- With the above change, we can probably also just get rid of the `ServiceCredentials` inner class.
- The whole `setCredentialFields()` method can also be deleted in favor of the new `parseCredentials()`.

**Comment by padamstx** [comment]
> I think the examples are missing the service url which is different than the authentication url

@germanattanasio Can you elaborate on what you think is missing?
The design wiki (https://github.ibm.com/CloudEngineering/openapi-sdkgen/wiki/Core-SDK-Authentication-Design) has examples on how to configure the service-level URL (and disable-ssl property) in a credentials file or environment variables.
The generated service class will have a default URL burned into it which would be used if the URL is set via external configuration.   This behavior is the same as it was prior to the authentication changes.

File: src/main/java/com/ibm/cloud/sdk/core/util/CredentialUtils.java (hunk 1 above)
Comment (padamstx) [reply src/main/java/com/ibm/cloud/sdk/core/util/CredentialUtils.java]: I was a little hesitant to remove all the "old" code from CredentialUtils as I wasn't sure how much (if any) we might have exposed to users.   But if you think that's not an issue, I'm happy to carve it up even more :)

File: src/main/java/com/ibm/cloud/sdk/core/security/ConfigBasedAuthenticatorFactory.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    Authenticator authenticator = null;
+
+    // If auth type was not specified, we'll use "iam" as the default.
+    String authType = props.get(Authenticator.PROPNAME_AUTH_TYPE);
+    if (StringUtils.isEmpty(authType)) {
+      authType = Authenticator.AUTHTYPE_IAM;
+    }
+
+    switch (authType) {
+    case Authenticator.AUTHTYPE_NOAUTH:
```
[the remark this answers is not in the document]
Comment (padamstx) [reply src/main/java/com/ibm/cloud/sdk/core/security/ConfigBasedAuthenticatorFactory.java]: done

File: src/main/java/com/ibm/cloud/sdk/core/security/CloudPakForDataAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+ */
+public class CloudPakForDataAuthenticator extends TokenRequestBasedAuthenticator<Cp4dToken, Cp4dTokenResponse>
+  implements Authenticator {
+
+  // This is the suffix we'll need to add to the user-supplied URL to retrieve an access token.
+  private static final String URL_SUFFIX = "/v1/preauth/validateAuth";
+
+  private String url;
+
+  // Hide the default ctor to force the use of the 1-arg ctor.
```
[the remark this answers is not in the document]
Comment (padamstx) [reply src/main/java/com/ibm/cloud/sdk/core/security/CloudPakForDataAuthenticator.java]: done

**Comment by padamstx** [comment]
@lpatino10 I've pushed an amended commit that addresses your comments.

**Pull request merged**